Note to self: 1 actual failing test (rest are network failures):

```
___________________________ TestReshaping.test_merge ___________________________
[gw1] linux -- Python 3.6.5 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.extension.json.test_json.TestReshaping object at 0x7fb92d208dd8>
data = JSONArary([{'D': 91, 'k': 84, 'L': 34, 'm': 63, 'C': 8, 'O': 1, 'R': 5}, {'T': 49, 'o': 46}, {'g': 81, 'O': 8, 'p': 54...': 32, 'p': 0, 'J': 89, 'd': 32, 's': 22}, {'W': 70, 'Y': 68, 'D': 87, 'k': 52}, {'F': 29, 'h': 10, 'S': 41, 'y': 24}])
na_value = {}
    def test_merge(self, data, na_value):
    
        df1 = pd.DataFrame({'int1': [1, 2, 3], 'key': [0, 1, 2],
                            'ext': data[:3]})
        df2 = pd.DataFrame({'int2': [1, 2, 3, 4], 'key': [0, 0, 1, 3]})
    
        res = pd.merge(df1, df2)
        exp = pd.DataFrame(
            {'int1': [1, 1, 2], 'int2': [1, 2, 3], 'key': [0, 0, 1],
             'ext': data._constructor_from_sequence(
                 [data[0], data[0], data[1]])})
>       self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
pandas/tests/extension/base/reshaping.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

E       AssertionError: DataFrame.columns are different
E       
E       DataFrame.columns values are different (75.0 %)
E       [left]:  Index(['int1', 'key', 'ext', 'int2'], dtype='object')
E       [right]: Index(['ext', 'int1', 'key', 'int2'], dtype='object')
```

Not seeing this locally, and also not failing for Decimal, so the column order seems to depend on the extension dtype? (which is a bit strange)

--- a/pandas/core/internals.py
+++ b/pandas/core/internals.py
@@ -5541,8 +5541,14 @@ def concatenate_join_units(join_units, concat_axis, copy):
     if len(to_concat) == 1:
         # Only one block, nothing to concatenate.
         concat_values = to_concat[0]
-        if copy and getattr(concat_values, 'base', 1) is not None:
-            concat_values = concat_values.copy()
+        if copy:
+            if isinstance(concat_values, np.ndarray):
+                # non-reindexed (=not yet copied) arrays are made into a view
+                # in JoinUnit.get_reindexed_values
+                if concat_values.base is not None:
+                    concat_values = concat_values.copy()
+            elif:
+                concat_values = concat_values.copy()
     else:
         concat_values = _concat._concat_compat(to_concat, axis=concat_axis)
 

--- a/pandas/tests/extension/base/reshaping.py
+++ b/pandas/tests/extension/base/reshaping.py
@@ -97,7 +97,7 @@ class BaseReshapingTests(BaseExtensionTests):
         assert df.dtypes['A'] == data.dtype
 
     def test_merge(self, data, na_value):
-
+        # GH-20743
         df1 = pd.DataFrame({'int1': [1, 2, 3], 'key': [0, 1, 2],
                             'ext': data[:3]})
         df2 = pd.DataFrame({'int2': [1, 2, 3, 4], 'key': [0, 0, 1, 3]})

--- a/pandas/tests/extension/base/reshaping.py
+++ b/pandas/tests/extension/base/reshaping.py
@@ -107,7 +107,7 @@ class BaseReshapingTests(BaseExtensionTests):
             {'int1': [1, 1, 2], 'int2': [1, 2, 3], 'key': [0, 0, 1],
              'ext': data._constructor_from_sequence(
                  [data[0], data[0], data[1]])})
-        self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
+        self.assert_frame_equal(res, exp[['int1', 'key', 'ext', 'int2']])
 
         res = pd.merge(df1, df2, how='outer')
         exp = pd.DataFrame(
@@ -115,4 +115,4 @@ class BaseReshapingTests(BaseExtensionTests):
              'key': [0, 0, 1, 2, 3],
              'ext': data._constructor_from_sequence(
                  [data[0], data[0], data[1], data[2], na_value])})
-        self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
+        self.assert_frame_equal(res, exp[['int1', 'key', 'ext', 'int2']])

--- a/pandas/tests/extension/decimal/test_decimal.py
+++ b/pandas/tests/extension/decimal/test_decimal.py
@@ -72,6 +72,14 @@ class BaseDecimal(object):
 
     def assert_frame_equal(self, left, right, *args, **kwargs):
         # TODO(EA): select_dtypes
+        tm.assert_index_equal(
+            left.columns, right.columns,
+            exact=kwargs.get('check_column_type', 'equiv'),
+            check_names=kwargs.get('check_names', True),
+            check_exact=kwargs.get('check_exact', False),
+            check_categorical=kwargs.get('check_categorical', True),
+            obj='{obj}.columns'.format(obj=kwargs.get('obj', 'DataFrame')))
+
         decimals = (left.dtypes == 'decimal').index
 
         for col in decimals:

